File: NextPermutation.py
from typing import List

class Solution:
    def nextPermutation(self, num: List[int]) -> None:
        n = len(num)
        i = n-2
        while i >= 0 and num[i+1] <= num[i]:
            i -=1

        if i >=0:
                j = n - 1

                while j > i and num[j] <= num[i]:
                    j -=1
                num[i], num[j] = num[j], num[i]
        num[i + 1:] = reversed(num[i + 1:])

File: Next Permutation/NextPermutation.java
/**
 * Time Complexity: O(n)
 * Space COmplexity: O(1)
 * Leetcode : Yes
 */

class Solution {
    public void nextPermutation(int[] nums) {
        int ptr1 = nums.length-1;
        // find non decreasing point
        while(ptr1 > 0 && nums[ptr1] <= nums[ptr1-1]) {            
                ptr1--;            
        }
        
        if(ptr1 > 0) {
            int ptr2 = nums.length-1;
            
            while(ptr2 > 0 && nums[ptr1-1] >= nums[ptr2]){
                ptr2--;
            }
            
            swap(nums, ptr1-1, ptr2);
            reverse(nums, ptr1, nums.length-1);
            
        }
        else {
            System.out.print("55");
            reverse(nums, 0, nums.length-1);
        }
        
    }
    
    public void reverse(int[] nums, int start, int end) {
        while(start < end) {
        swap(nums, start, end);
            start++;
            end--;
        }
    }
    
    public void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
    }
